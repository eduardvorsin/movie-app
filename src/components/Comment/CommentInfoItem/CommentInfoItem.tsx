'use client';
import { GeneralProps } from '@/types/shared';
import Link from 'next/link';
import { FocusEventHandler, MouseEventHandler, ReactNode } from 'react';

type Props = {
	children: ReactNode,
	type: 'author' | 'time',
	href?: string,
	onClick?: MouseEventHandler<HTMLSpanElement>,
	onFocus?: FocusEventHandler<HTMLSpanElement>,
} & GeneralProps;

export default function CommentInfoItem({
	type,
	className,
	children,
	onClick,
	onFocus,
	href,
	testId,
	...props
}: Props) {
	const classes = [
		'[&]:text-neutral-800 dark:[&]:text-dark-neutral-800 enabled:hover:[&]:text-neutral-700 dark:enabled:hover:[&]:text-dark-neutral-900 enabled:active:[&]:text-neutral-600 dark:enabled:active:[&]:text-dark-neutral-1000 transition-colors duration-150',
		type === 'author' ? 'font-medium' : '',
		className
	].join(' ');

	if (href) {
		return (
			<Link
				className={classes}
				onClick={onClick}
				onFocus={onFocus}
				href={href}
				data-testid={testId}
				prefetch
				{...props}
			>
				{children}
			</Link>
		);
	}

	return (
		<button
			className={classes}
			onClick={onClick}
			onFocus={onFocus}
			data-testid={testId}
			{...props}
		>
			{children}
		</button>
	);
};
