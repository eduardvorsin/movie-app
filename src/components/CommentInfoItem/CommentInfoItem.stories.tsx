import type { Meta, StoryObj } from '@storybook/react';
import CommentInfoItem from './CommentInfoItem';

const meta: Meta<typeof CommentInfoItem> = {
	title: 'components/CommentInfoItem',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes of the CommentInfoItem component'
		},
		children: {
			description: 'CommentInfoItem contents',
		},
		onFocus: {
			control: false,
			description: 'The callback is executed when the component gets focus'
		},
		onClick: {
			control: false,
			description: 'The callback is performed when the component is pressed'
		},
		type: {
			description: 'The type of element, there are 2 options available in total: author and time'
		},
		href: {
			description: 'Url for the link that will be clicked on'
		},
		testId: {
			description: 'Id for testing the component',
		}
	},
	component: CommentInfoItem,
	parameters: {
		docs: {
			description: {
				component: 'It is a component that allows you to add a block with the author\'s name or the time of publication of the comment. It is part of the Comment component.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof CommentInfoItem>;

export const Default: Story = {
	args: {
		children: 'Default'
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
};

export const Type: Story = {
	parameters: {
		docs: {
			description: {
				story: 'Allows you to set the block type. There are 2 types in total: author and time, which differ only in styling.',
			},
		},
	},
	render: () => (
		<div className='flex'>
			<CommentInfoItem
				className='mr-4 last:mr-0'
				type='author'
			>
				Author
			</CommentInfoItem>
			<CommentInfoItem
				className='mr-4 last:mr-0'
				type='time'
			>
				Time
			</CommentInfoItem>
		</div>
	),
};

export const EventCallbacks: Story = {
	argTypes: {
		onClick: {
			action: 'Clicked',
		},
		onFocus: {
			action: 'Focused',
		},
	},
	args: {
		children: 'click, focus, blur events',
		href: '/',
	},
	parameters: {
		docs: {
			description: {
				story: 'Callback functions can be supplied to onClick, onBlur, and onFocus events',
			},
		},
	},
};