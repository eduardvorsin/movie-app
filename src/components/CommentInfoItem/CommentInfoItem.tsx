import Link from 'next/link';
import { FocusEventHandler, MouseEventHandler, ReactNode } from 'react';

type Props = {
	children: ReactNode,
	type: 'author' | 'time',
	className?: string,
	href?: string,
	onClick?: MouseEventHandler<HTMLSpanElement>,
	onFocus?: FocusEventHandler<HTMLSpanElement>,
	testId?: string,
}

export default function CommentInfoItem({
	type,
	className,
	children,
	onClick,
	onFocus,
	href,
	testId,
}: Props) {
	const classes = [
		'text-neutral-800 dark:text-dark-neutral-800',
		type === 'author' ? 'font-medium' : '',
		className
	].join(' ');

	if (href) {
		return (
			<Link
				className={classes}
				onClick={onClick}
				onFocus={onFocus}
				href={href}
				data-testid={testId}
				prefetch
			>
				{children}
			</Link>
		);
	}

	return (
		<span
			className={classes}
			onClick={onClick}
			data-testid={testId}
		>
			{children}
		</span>
	);
};
