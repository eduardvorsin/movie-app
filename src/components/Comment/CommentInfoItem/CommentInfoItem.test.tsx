import { render, screen } from "@testing-library/react";
import CommentInfoItem from "./CommentInfoItem";
import userEvent from "@testing-library/user-event";

describe('CommentInfoItem tests', () => {
	it('is rendered correctly', () => {
		render(
			<CommentInfoItem
				type='author'
			>
				test item
			</CommentInfoItem>
		);

		expect(screen.getByRole<HTMLButtonElement>('button')).toBeInTheDocument();
	});

	it('if the prop href is not empty then the link is rendered', () => {
		render(
			<CommentInfoItem
				href='/'
				type='author'
			>
				test item
			</CommentInfoItem>
		);

		expect(screen.getByRole<HTMLAnchorElement>('link')).toBeInTheDocument();
	});

	it('when you click on an element, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();

		render(
			<CommentInfoItem
				href='/'
				type='author'
				onClick={mockFn}
			>
				test item
			</CommentInfoItem>
		);

		await user.click(screen.getByRole<HTMLAnchorElement>('link'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('when the focus is received, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();

		render(
			<CommentInfoItem
				href='/'
				type='author'
				onFocus={mockFn}
			>
				test item
			</CommentInfoItem>
		);

		await user.tab();

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('is a basic snapshot', () => {
		render(
			<CommentInfoItem
				type='author'
			>
				test item
			</CommentInfoItem>
		);

		expect(screen.getByRole<HTMLButtonElement>('button')).toMatchSnapshot();
	});

	it('is a snapshot with an element in the form of a link', () => {
		render(
			<CommentInfoItem
				type='author'
				href='/'
			>
				test item
			</CommentInfoItem>
		);

		expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
	});
});