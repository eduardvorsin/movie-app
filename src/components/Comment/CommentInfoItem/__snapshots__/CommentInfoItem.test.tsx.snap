// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentInfoItem tests is a basic snapshot 1`] = `
<button
  class="text-neutral-800 dark:text-dark-neutral-800 enabled:hover:text-neutral-700 dark:enabled:hover:text-dark-neutral-900 enabled:active:text-neutral-600 dark:enabled:active:text-dark-neutral-1000 font-medium "
>
  test item
</button>
`;

exports[`CommentInfoItem tests is a snapshot with an element in the form of a link 1`] = `
<a
  class="text-neutral-800 dark:text-dark-neutral-800 enabled:hover:text-neutral-700 dark:enabled:hover:text-dark-neutral-900 enabled:active:text-neutral-600 dark:enabled:active:text-dark-neutral-1000 font-medium "
  href="/"
>
  test item
</a>
`;
